src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.lang.stable;
+
+import jdk.internal.misc.Unsafe;
+import jdk.internal.vm.annotation.DontInline;
+import jdk.internal.vm.annotation.ForceInline;
+import jdk.internal.vm.annotation.Stable;
+
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+/**
+ * The implementation of StableValue.
+ *
+ * @implNote This implementation can be used early in the boot sequence as it does not
+ *           rely on reflection, MethodHandles, Streams etc.
+ *
+ * @param <T> type of the content
+ */
+public final class StableValueImpl<T> implements StableValue<T> {
+
+    // Unsafe allows StableValue to be used early in the boot sequence
+    static final Unsafe UNSAFE = Unsafe.getUnsafe();
+
+    // Unsafe offsets for direct field access
+    private static final long CONTENT_OFFSET =
+            UNSAFE.objectFieldOffset(StableValueImpl.class, "content");
+
+    // Generally, fields annotated with `@Stable` are accessed by the JVM using special
+    // memory semantics rules (see `parse.hpp` and `parse(1|2|3).cpp`).
+    //
+    // This field is used directly and reflectively via Unsafe using explicit memory semantics.
+    //
+    // | Value          |  Meaning      |
+    // | -------------- |  ------------ |
+    // | null           |  Unset        |
+    // | nullSentinel() |  Set(null)    |
+    // | other          |  Set(other)   |
+    //
+    @Stable
+    private Object content;
+
+    // Only allow creation via the factory `StableValueImpl::newInstance`
+    private StableValueImpl() {}
+
+    @ForceInline
+    @Override
+    public boolean trySet(T value) {
+        if (wrappedContentAcquire() != null) {
```

Comment:
IMHO, if our goal is to do:
```
Object content = this.content;
if (context != null) return content:
synchronized (...) {
    if (context != null) return content:
    this.context = ...
}
```

Then we might just use a volatile field and synchronized blocks. I don't see an immediate need for using acquire/release semantics -- especially when using a monitor. E.g. this should look more like a classic double checked locking idiom.